compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp

```diff
@@ -0,0 +1,931 @@
+  for (DistributionLayout *lattice : operands) {
+    if (lattice->hasLayout()) {
+      chosenOperandLayout = lattice;
+      break;
+    }
+  }
+
+  // Enforce the layout to other operands.
+  if (chosenOperandLayout) {
+    // Note that the operand lattice is not updated. So using the operand
+    // lattice again can cause bugs.
+    for (auto [index, lattice] : llvm::enumerate(operands)) {
+      OpOperand &opOperand = getOpOperand(op, index);
+      ChangeResult changed =
+          lattice->resolveWithPossibleConflict(chosenOperandLayout, opOperand);
+      update(lattice, changed);
+    }
+  }
+}
+
+/// ==========================================================================
+///        PROPAGATION TRANSFER FUNCTIONS
+/// ==========================================================================
+
+static void propagateLayoutToElementwiseOp(
+    Operation *op, ArrayRef<const DistributionLayout *> operandLattices,
+    ArrayRef<DistributionLayout *> resultLattices,
+    std::function<void(DistributionLayout *, ChangeResult)> update) {
+  // All operands and results must agree on the same layout.
+
+  // We do not support multiple results yet.
+  if (resultLattices.size() != 1)
+    return;
+
+  DistributionLayout *result = resultLattices[0];
+
+  // If result lattice already has a layout, we cannot do
+  // anything. We do not impose layout conflicts on results.
+  // TODO: Explore if this is actually needed.
+  if (result->hasLayout()) {
+    return;
+  }
+
+  // Check if all vector operands agree on the same layout.
+  const DistributionLayout *chosenOperandLayout =
+      getAgreedLayout(operandLattices);
+  if (chosenOperandLayout == nullptr) {
+    return;
+  }
+
+  ChangeResult changed = result->resolve(chosenOperandLayout);
+  update(result, changed);
+}
+
+static void propagateLayoutToMultiReductionOp(
+    vector::MultiDimReductionOp multiReduce,
+    ArrayRef<const DistributionLayout *> operandLattices,
+    ArrayRef<DistributionLayout *> resultLattices,
+    std::function<void(DistributionLayout *, ChangeResult)> update) {
+  // Multi reduce has only one vector result.
+  DistributionLayout *result = resultLattices[0];
+  // Multi reduce has first vector operands as the value being reduced.
+  const DistributionLayout *vector = operandLattices[0];
+  // Multi reduce has second operand as init.
+  const DistributionLayout *init = operandLattices[1];
+
+  // If result lattice already has a layout, we cannot do anything. We do not
+  // impose layout conflicts on results.
+  if (result->hasLayout()) {
+    return;
+  }
+
+  // If the vector begin reduced has a layout, then propagate it to the result.
+  // by projecting
+  if (vector->hasLayout()) {
+    SmallVector<bool> reductionMask = multiReduce.getReductionMask();
+    ChangeResult changed =
+        result->resolve(vector->getInnerLayout().project(reductionMask));
+    update(result, changed);
+    return;
+  }
+
+  // Otherwise, try resolving with init.
```

Comment:
What about checking the compatibility of the init value and the vector being reduced?

Reply:
They should be compatible by definition. Not sure if we need to check it?

Reply:
I think i understand what you mean now. I have a follow up patch in queue that adds that for this transfer function.

Reply:
thanks